File: 2370.longest-ideal-subsequence.cpp
/*
 * @lc app=leetcode id=2370 lang=cpp
 *
 * [2370] Longest Ideal Subsequence
 */

// @lc code=start
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int longestIdealString(string s, int k) {
        vector<int> dp(26, 0);
        for (char c : s) {
            int index = c - 'a';
            int curr = 0;
            for (int i = index - k; i <= index + k; i++) {
                if (0 <= i && i < 26) {
                    curr = max(curr, dp[i] + 1);
                }
            }
            dp[index] = curr;
        }
        int maxCount = 0;
        for (int count : dp) {
            maxCount = max(maxCount, count);
        }
        return maxCount;
    }
};

// @lc code=end

File: 2370.longest-ideal-subsequence.py
#
# @lc app=leetcode id=2370 lang=python3
#
# [2370] Longest Ideal Subsequence
#

# @lc code=start
class Solution:
    def longestIdealString(self, s: str, k: int) -> int:
        dp = [0] * 26
        for c in s:
            index = ord(c) - ord('a')
            curr = 0
            for i in range(index - k, index + k + 1, 1):
                if 0 <= i < 26: curr = max(curr, dp[i] + 1)
            dp[index] = curr
        return max(dp)
# @lc code=end

